text/0569-tracked-list.md

```diff
@@ -0,0 +1,306 @@
+
+### `Ember.A`
+
+`Ember.A` will do nothing if it receives a `TrackedList`, and return the
+original value instead.
+
+## How we teach this
+
+`TrackedList` will be the default reactive array class recommended moving
+forward, and is meant to replace `EmberArray` entirely. The main guides should
+be updated to reflect this. In the advanced guides, there should be a section
+covering the functional-immutable style of updating arrays and objects.
+
+### API Docs
+
+TODO
+
+## Drawbacks
+
+- Introduces a new array-like class, which may cause confusion as the community
+  transitions.
+
+- Introduces an array class that _may_ have a limited shelf life. IE11 support
+  will eventually be removed, and allow us to use native `Proxy` to
+  transparently wrap real native arrays instead. However, this potential class
+  would have a few of its own caveats:
+  - Performance. Proxies have gotten much better, but they are still much less
+    performant compared to native arrays.
+  - General interop concerns. Arrays have a lot of odd behaviors that have been
+    built up over the years, such as the behaviors of sparse arrays, the
+    differences between how numeric and string based keys are handled, etc.
+  Given these caveats, and the fact that IE11 is still supported and will be for
+  the indeterminate future, a minimal array-like class that improves on the
+  existing DX seems like a good middle-ground solution for the time being.
+
+## Alternatives
+
+- `getAt` and `setAt` could be changed to match the `objectAt` and `replace`
+  methods that exist on `EmberArray`. This would be better for interoperability,
+  but comes with caveats:
+  - `objectAt` is not accurate for values that contain primitive values
+  - `replace` is a not an ideal method for updating a value at a predetermined
+    index. It's API is more similar to `splice`, which is confusing and
+    difficult to work with.
+  Some potential alternatives include:
+  - `objectAt` and `replaceAt`
+  - `valueAt` and `replaceAt`
+  - `valueAt` and `setValueAt`
+  `TrackedList` could also implement `objectAt` and `replace` in order to remain
+  mostly compatible with legacy code, but recommend that `getAt`/`setAt` be used
+  by convention.
+
+- Passing the underlying array as the third argument to methods like `map` and
+  `forEach` may be confusing to users. However, passing the list itself instead
+  would remove many of the browsers optimizations for these methods. They are
+  implemented in most modern runtimes as precompiled, pre-JITed code, and
+  matching anything near the level of native performance is likely impossible.
+  The other alternative would be to not pass a third argument at all (in
+  development mode), which would prevent users from relying on it in the first
+  place. This would limit the interoperability of `TrackedList` somewhat, but
+  would likely not have a massive impact.
+
+## Unresolved questions
```

Comment:
Is `TrackedArray` to be subclassed? If so, what invariants do subclasses need to uphold?

Reply:
I had exactly this question.

Also, what is the intended API for a subclass to manually invalidate itself?

Reply:
`TrackedList` is not intended to be subclassed, though there isn't a way to prevent it in JS, and there is no API for manual invalidation. If users want to implement a custom iterable, the recommendation would instead be to use the standard `Symbol.iterable` API to do so, as it should be fully supported by Ember at this point. If there are still some gaps, they should probably be addressed in a separate RFC.